File: go/0209M-长度最小的子数组.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-14 10:18:21
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func minSubArrayLen(target int, nums []int) int {
	l := len(nums)
	ans, sum, k := l+1, 0, 0
	var sub = 0 // 滑动窗口长度
	for i := 0; i < l; i++ {
		sum += nums[i]
		for sum >= target {
			sub = i - k + 1
			if sub < ans {
				ans = sub
			}
			sum -= nums[k]
			k++
		}
	}
	if ans == l+1 {
		return 0
	} else {
		return ans
	}
}

File: cpp/0209M-长度最小的子数组.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-13 09:18:16
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

using namespace std;

class Solution {
   public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int ans = INT32_MAX;
        int sum = 0;
        for (int l = 0, r = 0; r < nums.size(); r++) {
            sum += nums[r];
            while (sum >= target) {
                ans = min(ans, r - l + 1);
                sum -= nums[l++];
            }
        }
        if (ans == INT32_MAX) return 0;
        return ans;
    }
};
